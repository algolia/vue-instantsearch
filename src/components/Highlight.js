import { getPropertyByPath } from '../util/object';
import { warn } from '../util/warn';

export default {
  functional: true,
  props: {
    result: {
      type: Object,
      required: true,
    },
    attributeName: {
      type: String,
      required: true,
    },
  },
  render(h, ctx) {
    const result = ctx.props.result;
    const attributeName = ctx.props.attributeName;

    const attributePath = `_highlightResult.${attributeName}.value`;
    const attributeValue = getPropertyByPath(result, attributePath);

    if (process.env.NODE_ENV !== 'production' && attributeValue === undefined) {
      warn(
        `The "${attributeName}" attribute might currently not be configured to be highlighted in Algolia.
        See https://www.algolia.com/doc/api-reference/api-parameters/attributesToHighlight/.`
      );
    }

    return h('span', {
      class: {
        'ais-highlight': true,
      },
      domProps: {
        innerHTML: attributeValue,
      },
    });
  },
};
