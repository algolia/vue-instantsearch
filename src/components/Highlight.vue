<template>
  <span
    :class="suit()"
    v-html="innerHTML"
  />
</template>

<script>
import instantsearch from 'instantsearch.js/es';
import { createSuitMixin } from '../mixins/suit';

export default {
  name: 'AisHighlight',
  mixins: [createSuitMixin({ name: 'Highlight' })],
  props: {
    hit: {
      type: Object,
      required: true,
    },
    attribute: {
      type: String,
      required: true,
    },
    highlightedTagName: {
      type: String,
      default: 'mark',
    },
  },
  computed: {
    innerHTML() {
      return instantsearch.highlight({
        attribute: this.attribute,
        hit: this.hit,
        highlightedTagName: this.highlightedTagName,
      });
    },
  },
};
</script>
