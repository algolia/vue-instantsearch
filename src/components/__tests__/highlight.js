import { mount } from '@vue/test-utils';
import Highlight from '../Highlight.vue';

afterEach(() => {
  process.env.NODE_ENV = 'test';
});

test('renders proper HTML', () => {
  const hit = {
    _highlightResult: {
      attr: {
        value: `con<em>ten</em>t`,
      },
    },
  };

  const wrapper = mount(Highlight, {
    propsData: {
      attribute: 'attr',
      hit,
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('renders proper HTML with highlightTagName', () => {
  const hit = {
    _highlightResult: {
      attr: {
        value: `con<em>ten</em>t`,
      },
    },
  };

  const wrapper = mount(Highlight, {
    propsData: {
      attribute: 'attr',
      highlightedTagName: 'marquee',
      hit,
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('should render an empty string in production if attribute is not highlighted', () => {
  process.env.NODE_ENV = 'production';
  const hit = {
    _highlightResult: {},
  };

  const wrapper = mount(Highlight, {
    propsData: {
      attribute: 'attr',
      hit,
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('should warn when not in production if attribute is not highlighted', () => {
  global.console.warn = jest.fn();

  const hit = {
    _highlightResult: {},
  };

  mount(Highlight, {
    propsData: {
      attribute: 'attr',
      hit,
    },
  });

  expect(global.console.warn).toHaveBeenCalledTimes(1);
});

test('allows usage of dot delimited path to access nested attribute', () => {
  const hit = {
    _highlightResult: {
      attr: {
        nested: {
          value: `nested <em>val</em>`,
        },
      },
    },
  };

  const wrapper = mount(Highlight, {
    propsData: {
      attribute: 'attr.nested',
      hit,
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});
