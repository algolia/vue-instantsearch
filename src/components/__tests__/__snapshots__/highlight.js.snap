// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows usage of dot delimited path to access nested attribute 1`] = `

<span class="ais-Highlight">
  nested
  <mark class="ais-Highlight-highlighted">
    val
  </mark>
</span>

`;

exports[`renders proper HTML 1`] = `

<span class="ais-Highlight">
  con
  <mark class="ais-Highlight-highlighted">
    ten
  </mark>
  t
</span>

`;

exports[`renders proper HTML with highlightTagName 1`] = `

<span class="ais-Highlight">
  con
  <marquee class="ais-Highlight-highlighted">
    ten
  </marquee>
  t
</span>

`;

exports[`should render an empty string in production if attribute is not highlighted 1`] = `

<span class="ais-Highlight">
</span>

`;
